File: ODataDataProvider/ODataVirtualDataSourceDataProvider.cs
using Infragistics.Controls;
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Linq;
using System.Net;
using System.Text;
using System.Threading.Tasks;
using Simple.OData.Client;
using System.Diagnostics;
#if !PCL
using Infragistics.Controls.DataSource;
#endif

#if PCL
using Infragistics.Core.Controls.DataSource;
#endif

#if DATA_PRESENTER
namespace Reference.DataSources.OData
#else
namespace Infragistics.Controls.DataSource
#endif
{
    public class ODataVirtualDataSourceDataProvider
        : IDataSourceVirtualDataProvider
    {

        private ODataVirtualDataSourceDataProviderWorker _worker;
        private LinkedList<int> _requests = new LinkedList<int>();
        private DataSourcePageLoadedCallback _callback;

        public ODataVirtualDataSourceDataProvider()
        {
            _sortDescriptions = new SortDescriptionCollection();
            _groupDescriptions = new SortDescriptionCollection();
            _sortDescriptions.CollectionChanged += SortDescriptions_CollectionChanged;
            _groupDescriptions.CollectionChanged += GroupDescriptions_CollectionChanged;
            _filterExpressions = new FilterExpressionCollection();
            _filterExpressions.CollectionChanged += FilterExpressions_CollectionChanged;
            _summaryDescriptions = new SummaryDescriptionCollection();
            _summaryDescriptions.CollectionChanged += SummaryDescriptions_CollectionChanged;
        }

        private void FilterExpressions_CollectionChanged(object sender, System.Collections.Specialized.NotifyCollectionChangedEventArgs e)
        {
            QueueAutoRefresh();
        }

        private void SortDescriptions_CollectionChanged(object sender, System.Collections.Specialized.NotifyCollectionChangedEventArgs e)
        {
            QueueAutoRefresh();
        }

        private void GroupDescriptions_CollectionChanged(object sender, System.Collections.Specialized.NotifyCollectionChangedEventArgs e)
        {
            QueueAutoRefresh();
        }

        private void SummaryDescriptions_CollectionChanged(object sender, System.Collections.Specialized.NotifyCollectionChangedEventArgs e)
        {
            QueueAutoRefresh();
        }

        public void AddPageRequest(int pageIndex, DataSourcePageRequestPriority priority)
        {
            if (DeferAutoRefresh)
            {
                return;
            }

            if (_worker != null && _worker.IsShutdown)
            {
                _worker = null;
                _callback = null;
            }

            if (_worker == null)
            {
                CreateWorker();
            }

            if (priority == DataSourcePageRequestPriority.High)
            {
                _requests.AddFirst(pageIndex);
            }
            else
            {
                _requests.AddLast(pageIndex);
            }
            if (!_worker.AddPageRequest(pageIndex, priority))
            {
                _worker = null;
                _callback = null;
                AddPageRequest(pageIndex, priority);
            }
        }

        private void CreateWorker()
        {
            if (!Valid())
            {
                return;
            }

            _callback = RaisePageLoaded;

            var settings = GetWorkerSettings();

            _worker = new ODataVirtualDataSourceDataProviderWorker(
                settings);
        }

        private bool Valid()
        {
            return EntitySet != null &&
                BaseUri != null;
        }

        private ODataVirtualDataSourceDataProviderWorkerSettings GetWorkerSettings()
        {
            return new ODataVirtualDataSourceDataProviderWorkerSettings()
            {
                BaseUri = _baseUri,
                EntitySet = _entitySet,
                PageSizeRequested = _pageSizeRequested,
                TimeoutMilliseconds = _timeoutMilliseconds,
                PageLoaded = _callback,
                ExecutionContext = _executionContext,
                SortDescriptions = _sortDescriptions,
                FilterExpressions = _filterExpressions,
                PropertiesRequested = _propertiesRequested,
                GroupDescriptions = _groupDescriptions,
                IsAggregationSupportedByServer = _isAggregationSupportedByServer,
                SummaryDescriptions = _summaryDescriptions,
                SummaryScope = _summaryScope
            };
        }

        public void RemovePageRequest(int pageIndex)
        {
            _requests.Remove(pageIndex);
            if (_worker == null)
            {
                return;
            }
            _worker.RemovePageRequest(pageIndex);
        }

        public void RemoveAllPageRequests()
        {
            _requests.Clear();
            if (_worker == null)
            {
                return;
            }
            _worker.RemoveAllPageRequests();
        }

        public void Close()
        {
            if (_worker != null)
            {
                _worker.Shutdown();
                _worker = null;
                _callback = null;
            }
        }

        private DataSourcePageLoadedCallback _pageLoaded;
        public DataSourcePageLoadedCallback PageLoaded
        {
            get
            {
                return _pageLoaded;
            }
            set
            {
                _pageLoaded = value;
                QueueAutoRefresh();
            }
        }

        private void RaisePageLoaded(IDataSourcePage page, int fullCount, int actualPageSize)
        {
            if (_pageLoaded != null)
            {
                _currentFullCount = fullCount;
                if (_currentSchema == null)
                {
                    IDataSourceSchema currentSchema = null;
                    if (page != null)
                    {
                        currentSchema = page.Schema();
                    }

                    _currentSchema = currentSchema;
                    if (SchemaChanged != null)
                    {
                        SchemaChanged(this, new DataSourceDataProviderSchemaChangedEventArgs(_currentSchema, _currentFullCount));
                    }
                }

                if (page.PageIndex() != ODataVirtualDataSourceDataProviderWorker.SchemaRequestIndex)
                {
                    _pageLoaded(page, fullCount, actualPageSize);
                }
            }
        }

        private void KillWorker()
        {
            if (_worker != null)
            {
                _worker.Shutdown();
                _worker = null;
                _callback = null;
            }
        }

        private int _pageSizeRequested = 50;
        public int PageSizeRequested
        {
            get
            {
                return _pageSizeRequested;
            }
            set
            {
                _pageSizeRequested = value;
                QueueAutoRefresh();
            }
        }

        private string _baseUri = null;
        public string BaseUri
        {
            get
            {
                return _baseUri;
            }
            set
            {
                var oldValue = _baseUri;
                _baseUri = value;
                if (oldValue != _baseUri)
                {
                    QueueAutoRefresh();
                    if (Valid() && DeferAutoRefresh)
                    {
                        QueueSchemaFetch();
                    }
                }
            }
        }

        private string _entitySet = null;
        public string EntitySet
        {
            get
            {
                return _entitySet;
            }
            set
            {
                var oldValue = _entitySet;
                _entitySet = value;
                if (oldValue != _entitySet)
                {
                    QueueAutoRefresh();
                    if (Valid() && DeferAutoRefresh)
                    {
                        QueueSchemaFetch();
                    }
                }
            }
        }

        private int _timeoutMilliseconds = 10000;
        public int TimeoutMilliseconds
        {
            get
            {
                return _timeoutMilliseconds;
            }
            set
            {
                _timeoutMilliseconds = value;
                QueueAutoRefresh();
            }
        }

        private bool _isAggregationSupportedByServer = false;
        /// <summary>
        /// Gets or sets whether the server supports aggregation query options required for grouping.
        /// </summary>
        public bool IsAggregationSupportedByServer
        {
            get
            {
                return _isAggregationSupportedByServer;
            }
            set
            {
                var oldValue = _isAggregationSupportedByServer;
                _isAggregationSupportedByServer = value;
                QueueAutoRefresh();
            }
        }

        public object GetItemValue(object item, string valueName)
        {
            var dic = (IDictionary<string, object>)item;
            if (dic.ContainsKey(valueName))
            {
                return dic[valueName];
            }
            else
            {
                return null; 
            }
        }

        public event DataSourceDataProviderSchemaChangedHandler SchemaChanged;
        private int _currentFullCount = 0;
        private IDataSourceSchema _currentSchema;

        public int ActualCount
        {
            get
            {
                return _currentFullCount;
            }
        }

        public IDataSourceSchema ActualSchema
        {
            get
            {
                return _currentSchema;
            }
        }

        private IDataSourceExecutionContext _executionContext;
        public IDataSourceExecutionContext ExecutionContext
        {
            get
            {
                return _executionContext;
            }
            set
            {
                _executionContext = value;
                QueueAutoRefresh();
            }
        }

        private IDataSourceDataProviderUpdateNotifier _updateNotifier;
        public IDataSourceDataProviderUpdateNotifier UpdateNotifier
        {
            get
            {
                return _updateNotifier;
            }
            set
            {
                _updateNotifier = value;
            }
        }

        private bool _deferAutoRefresh = false;
        public bool DeferAutoRefresh
        {
            get
            {
                return _deferAutoRefresh;
            }

            set
            {
                _deferAutoRefresh = value;
                if (!_deferAutoRefresh)
                {
                    QueueAutoRefresh();
                }
                if (_deferAutoRefresh && Valid() && _currentSchema == null)
                {
                    QueueSchemaFetch();
                }
            }
        }

        public bool IsSortingSupported
        {
            get
            {
                return true;
            }
        }

        public bool IsFilteringSupported
        {
            get
            {
                return true;
            }
        }

        private SortDescriptionCollection _sortDescriptions;
        public SortDescriptionCollection SortDescriptions
        {
            get
            {
                return _sortDescriptions;
            }
        }

        private string[] _propertiesRequested;
        public string[] PropertiesRequested
        {
            get
            {
                return _propertiesRequested;
            }
            set
            {
                _propertiesRequested = value;
                QueueAutoRefresh();
            }
        }

        private FilterExpressionCollection _filterExpressions;
        public FilterExpressionCollection FilterExpressions
        {
            get
            {
                return _filterExpressions;
            }
        }

        public bool NotifyUsingSourceIndexes
        {
            get
            {
                return true;
            }
        }

        public bool IsItemIndexLookupSupported
        {
            get
            {
                return false;
            }
        }

        public bool IsKeyIndexLookupSupported
        {
            get
            {
                return false;
            }
        }

        public bool IsGroupingSupported
        {
            get
            {
                return false;
            }
        }

        private SortDescriptionCollection _groupDescriptions;
        public SortDescriptionCollection GroupDescriptions
        {
            get
            {
                return _groupDescriptions;
            }
        }

        private SummaryDescriptionCollection _summaryDescriptions;
        public SummaryDescriptionCollection SummaryDescriptions
        {
            get
            {
                return _summaryDescriptions;
            }
        }

        private DataSourceSummaryScope _summaryScope;
        public DataSourceSummaryScope SummaryScope
        {
            get
            {
                return _summaryScope;
            }
            set
            {
                _summaryScope = value;
            }
        }

        public void NotifySetItem(int index, object oldItem, object newItem)
        {
            if (UpdateNotifier != null)
            {
                UpdateNotifier.NotifySetItem(index, oldItem, newItem);
            }
        }

        public void NotifyClearItems()
        {
            if (UpdateNotifier != null)
            {
                UpdateNotifier.NotifyClearItems();
            }
        }

        public void NotifyInsertItem(int index, object newItem)
        {
            if (UpdateNotifier != null)
            {
                UpdateNotifier.NotifyInsertItem(index, newItem);
            }
        }

        public void NotifyRemoveItem(int index, object oldItem)
        {
            if (UpdateNotifier != null)
            {
                UpdateNotifier.NotifyRemoveItem(index, oldItem);
            }
        }

        internal bool _schemaFetchQueued = false;
        public void QueueSchemaFetch()
        {
            if (_schemaFetchQueued)
            {
                return;
            }

            if (ExecutionContext != null)
            {
                _schemaFetchQueued = true;
                ExecutionContext.EnqueueAction(DoSchemaFetchInternal);
            }
        }

        internal void DoSchemaFetchInternal()
        {
            if (!_schemaFetchQueued)
            {
                return;
            }
            _schemaFetchQueued = false;
            SchemaFetchInternal();
        }

        internal void SchemaFetchInternal()
        {
            SchemaFetchInternalOverride();
        }

        protected virtual void SchemaFetchInternalOverride()
        {
            if (!DeferAutoRefresh)
            {
                return;
            }
            RemoveAllPageRequests();
            KillWorker();
            CreateWorker();
        
            AddSchemaRequest();
        }

        private void AddSchemaRequest()
        {
            _worker.AddPageRequest(
                ODataVirtualDataSourceDataProviderWorker.SchemaRequestIndex,
                DataSourcePageRequestPriority.High
                );
        }

        internal bool _autoRefreshQueued = false;
        public void QueueAutoRefresh()
        {
            if (DeferAutoRefresh)
            {
                return;
            }
            if (_autoRefreshQueued)
            {
                return;
            }

            if (ExecutionContext != null)
            {
                _autoRefreshQueued = true;
                ExecutionContext.EnqueueAction(DoRefreshInternal);
            }
        }

        internal void DoRefreshInternal()
        {
            if (DeferAutoRefresh)
            {
                _autoRefreshQueued = false;
                return;
            }
            if (!_autoRefreshQueued)
            {
                return;
            }
            _autoRefreshQueued = false;
            RefreshInternal();
        }

        internal void RefreshInternal()
        {
            RefreshInternalOverride();
        }

        protected virtual void RefreshInternalOverride()
        {
            RemoveAllPageRequests();
            KillWorker();
            CreateWorker();
            //TODO: should we have this here? prob need to readd request for current page instead.
            _worker.AddPageRequest(0, DataSourcePageRequestPriority.Normal);
        }

        public void FlushAutoRefresh()
        {
            DoRefreshInternal();
        }

        public void Refresh()
        {
            RefreshInternal();
        }

        public int IndexOfItem(object item)
        {
            return -1;
        }

        public int IndexOfKey(object[] key)
        {
            return -1;
        }
        
        public DataSourceSchemaPropertyType ResolveSchemaPropertyType(string propertyPath)
		{
			var propertiesInPath = propertyPath.Split('.');
			if (propertiesInPath.Length == 1 && this.ActualSchema != null)
			{
				for (int i = 0; i < this.ActualSchema.PropertyNames.Length; i++)
				{
					if (this.ActualSchema.PropertyNames[i] == propertiesInPath[0])
					{
						return this.ActualSchema.PropertyTypes[i];
					}
				}
			}

			return DataSourceSchemaPropertyType.ObjectValue;
		}

        public IDataSourceDataProvider Clone()
        {
            var provider = new ODataVirtualDataSourceDataProvider();
            provider.ExecutionContext = ExecutionContext;
            provider.EntitySet = EntitySet;
            provider.PageSizeRequested = PageSizeRequested;
            provider.PropertiesRequested = PropertiesRequested;
            provider.SummaryScope = SummaryScope;
            provider.TimeoutMilliseconds = TimeoutMilliseconds;

            for (var i = 0; i < FilterExpressions.Count; i++)
                provider.FilterExpressions.Add(FilterExpressions[i]);

            for (var i = 0; i < SortDescriptions.Count; i++)
                provider.SortDescriptions.Add(SortDescriptions[i]);

            for (var i = 0; i < GroupDescriptions.Count; i++)
                provider.GroupDescriptions.Add(GroupDescriptions[i]);

            for (var i = 0; i < SummaryDescriptions.Count; i++)
                provider.SummaryDescriptions.Add(SummaryDescriptions[i]);

            return provider;
        }
    }

   
}

File: AndroidOData/odatavirtualdatasource/src/main/java/com/infragistics/odatavirtualdatasource/ODataVirtualDataSourceDataProvider.java
package com.infragistics.odatavirtualdatasource;

import com.fasterxml.jackson.databind.util.LinkedNode;
import com.infragistics.controls.*;
import org.apache.olingo.client.api.domain.ClientEntity;
import org.apache.olingo.client.api.domain.ClientValue;

import java.util.*;

public class ODataVirtualDataSourceDataProvider
    implements DataSourceVirtualDataProvider {

    private ODataVirtualDataSourceDataProviderWorker _worker;
    private LinkedList<Integer> _requests = new LinkedList<Integer>();
    private DataSourcePageLoadedCallback _callback;
    private SortDescriptionCollection _sortDescriptions = new SortDescriptionCollection();
    private FilterExpressionCollection _filterExpressions = new FilterExpressionCollection();

    public ODataVirtualDataSourceDataProvider() {
        _sortDescriptions.addChangedListener(new SyncableObservableCollectionChangedListener() {
            @Override
            public void onChanged(Object dataSourceSortDescriptionCollection) {
                onSortDescriptionsChanged();
            }
        });
    }

    public boolean getIsSortingSupported() {
        return true;
    }

    public SortDescriptionCollection getSortDescriptions() {
        return _sortDescriptions;
    }

    public boolean getIsFilteringSupported() {
        return true;
    }

    public FilterExpressionCollection getFilterExpressions() {
        return _filterExpressions;
    }

    private String[] _propertiesRequested;
    @Override
    public String[] setPropertiesRequested(String[] strings) {
        _propertiesRequested = strings;

        queueAutoRefresh();
        return _propertiesRequested;
    }

    @Override
    public String[] getPropertiesRequested() {
        return _propertiesRequested;
    }

    private void onSortDescriptionsChanged() {
        queueAutoRefresh();
    }

    @Override
    public void addPageRequest(int pageIndex, DataSourcePageRequestPriority dataSourcePageRequestPriority) {
        if (getDeferAutoRefresh()) {
            return;
        }
        if (_worker != null && _worker.getIsShutdown()) {
            _worker = null;
            _callback = null;
        }

        if (_worker == null)
        {
            createWorker();
        }

        if (dataSourcePageRequestPriority == DataSourcePageRequestPriority.HIGH)
        {
            _requests.addFirst(pageIndex);
        }
        else
        {
            _requests.addLast(pageIndex);
        }
        if (!_worker.addPageRequest(pageIndex, dataSourcePageRequestPriority)) {
            _worker = null;
            _callback = null;
            addPageRequest(pageIndex, dataSourcePageRequestPriority);
        }
    }

    private void createWorker() {
        _callback = new DataSourcePageLoadedCallback() {
            @Override
            public void invoke(DataSourcePage dataSourcePage, int fullCount, int actualPageSize) {
                raisePageLoaded(dataSourcePage, fullCount, actualPageSize);
            }
        };

        ODataVirtualDataSourceDataProviderWorkerSettings settings = new ODataVirtualDataSourceDataProviderWorkerSettings();
        settings.setBsaeUri(_baseURI);
        settings.setEntitySet(_entitySet);
        settings.setPageSizeRequested(_pageSizeRequested);
        settings.setTimeoutMilliseconds(_timeoutMilliseconds);
        settings.setMetadataType(_metadataType);
        settings.setPageLoaded(_callback);
        settings.setExecutionContext(_executionContext);
        settings.setSortDescriptions(_sortDescriptions);
        settings.setFilterExpressions(_filterExpressions);
        settings.setDesiredProperties(_propertiesRequested);

        _worker = new ODataVirtualDataSourceDataProviderWorker(settings);
    }

    @Override
    public void removePageRequest(int pageIndex) {
        _requests.removeFirstOccurrence(pageIndex);
        if (_worker == null)
        {
            return;
        }
        _worker.removePageRequest(pageIndex);
    }

    @Override
    public void removeAllPageRequests() {
        _requests.clear();
        if (_worker == null)
        {
            return;
        }
        _worker.removeAllPageRequests();
    }

    @Override
    public void close() {
        if (_worker != null)
        {
            _worker.shutdown();
            _worker = null;
            _callback = null;
        }
    }

    private DataSourcePageLoadedCallback _pageLoaded;
    @Override
    public DataSourcePageLoadedCallback setPageLoaded(DataSourcePageLoadedCallback pageLoaded) {
        _pageLoaded = pageLoaded;
        killWorker();
        queueAutoRefresh();
        return _pageLoaded;
    }

    private boolean valid()
    {
        return _baseURI != null &&
                _entitySet != null;
    }

    @Override
    public DataSourcePageLoadedCallback getPageLoaded() {
        return _pageLoaded;
    }

    private void raisePageLoaded(DataSourcePage page, int fullCount, int actualPageSize)
    {
        if (_pageLoaded != null)
        {
            _currentFullCount = fullCount;

            if (_currentSchema == null)
            {
                if (page != null)
                {
                    _currentSchema = page.schema();
                }
                for (int i = 0; i < _initializedHandlers.size(); i++) {
                    OnDataSourceDataProviderSchemaChangedListener handler = _initializedHandlers.get(i);
                    if (handler != null) {
                        handler.onSchemaChanged(this, new DataSourceDataProviderSchemaChangedEvent(_currentSchema, _currentFullCount));
                    }
                }
            }
            if (page.pageIndex() != ODataVirtualDataSourceDataProviderWorker.SCHEMA_REQUEST_INDEX) {
                _pageLoaded.invoke(page, fullCount, actualPageSize);
            }
        }
    }

    private void killWorker()
    {
        if (_worker != null) {
            _worker.shutdown();
            _worker = null;
            _callback = null;
        }
    }

    private int _pageSizeRequested = 50;
    @Override
    public int setPageSizeRequested(int value) {
        _pageSizeRequested = value;
        killWorker();
        queueAutoRefresh();
        return _pageSizeRequested;
    }

    @Override
    public int getPageSizeRequested() {
        return _pageSizeRequested;
    }

    private int _timeoutMilliseconds = 10000;
    public int setTimeoutMilliseconds(int value) {
        _timeoutMilliseconds = value;
        killWorker();
        queueAutoRefresh();
        return _timeoutMilliseconds;
    }

    public int getTimeoutMilliseconds() {
        return _timeoutMilliseconds;
    }

    private String _baseURI = null;
    public String setBaseURI(String value) {
        String oldValue = _baseURI;
        _baseURI = value;
        if (oldValue != _baseURI) {
            queueAutoRefresh();
            if (valid() && getDeferAutoRefresh()) {
                queueSchemaFetch();
            }
        }
        return value;
    }

    private ODataVirtualDataSourceMetadataType _metadataType = ODataVirtualDataSourceMetadataType.MINIMAL;
    public ODataVirtualDataSourceMetadataType setMetadataType(ODataVirtualDataSourceMetadataType value) {
        _metadataType = value;
        killWorker();
        queueAutoRefresh();
        return _metadataType;
    }

    public String getBaseURI() {
        return _baseURI;
    }

    private String _entitySet = null;
    public String setEntitySet(String value) {
        String oldValue = _entitySet;
        _entitySet = value;
        if (oldValue != _entitySet) {
            queueAutoRefresh();
            if (valid() && getDeferAutoRefresh()) {
                queueSchemaFetch();
            }
        }
        return value;
    }

    public String getEntitySet() {
        return _entitySet;
    }

    @Override
    public Object getItemValue(Object o, String s) {
        ClientEntity entity = (ClientEntity)o;
        ClientValue value = entity.getProperty(s).getValue();
        if (!value.isPrimitive()) {
            return value;
        }
        return value.asPrimitive().toValue();
    }

    private List<OnDataSourceDataProviderSchemaChangedListener> _initializedHandlers = new ArrayList<OnDataSourceDataProviderSchemaChangedListener>();
    @Override
    public void addOnSchemaChangedListener(OnDataSourceDataProviderSchemaChangedListener onDataSourceDataProviderSchemaChangedListener) {
        _initializedHandlers.add(onDataSourceDataProviderSchemaChangedListener);
    }

    @Override
    public void removeOnSchemaChangedListener(OnDataSourceDataProviderSchemaChangedListener onDataSourceDataProviderSchemaChangedListener) {
        _initializedHandlers.remove(onDataSourceDataProviderSchemaChangedListener);
    }

    private int _currentFullCount = 0;
    private DataSourceSchema _currentSchema;

    @Override
    public int getActualCount() {
        return _currentFullCount;
    }

    @Override
    public DataSourceSchema getActualSchema() {
        return _currentSchema;
    }

    private DataSourceExecutionContext _executionContext;
    @Override
    public DataSourceExecutionContext setExecutionContext(DataSourceExecutionContext value) {
        _executionContext = value;
        killWorker();
        queueAutoRefresh();
        return value;
    }

    @Override
    public DataSourceExecutionContext getExecutionContext() {
        return _executionContext;
    }

    private DataSourceDataProviderUpdateNotifier _updateNotifier;
    @Override
    public DataSourceDataProviderUpdateNotifier setUpdateNotifier(DataSourceDataProviderUpdateNotifier dataSourceDataProviderUpdateNotifier) {
        _updateNotifier = dataSourceDataProviderUpdateNotifier;
        return _updateNotifier;
    }

    @Override
    public DataSourceDataProviderUpdateNotifier getUpdateNotifier() {
        return _updateNotifier;
    }

    @Override
    public void notifySetItem(int i, Object oldItem, Object newItem) {
        if (_updateNotifier != null) {
            _updateNotifier.notifySetItem(i, oldItem, newItem);
        }
    }

    @Override
    public void notifyClearItems() {
        if (_updateNotifier != null) {
            _updateNotifier.notifyClearItems();
        }
    }

    @Override
    public void notifyInsertItem(int i, Object newItem) {
        if (_updateNotifier != null) {
            _updateNotifier.notifyInsertItem(i, newItem);
        }
    }

    @Override
    public void notifyRemoveItem(int i, Object oldItem) {
        if (_updateNotifier != null) {
            _updateNotifier.notifyRemoveItem(i, oldItem);
        }
    }

    public boolean _autoRefreshQueued = false;


    public void queueAutoRefresh() {
        if (getDeferAutoRefresh()) {
            return;
        }

       if (_autoRefreshQueued) {
            return;
        }

        if (getExecutionContext() != null) {
            _autoRefreshQueued = true;
            getExecutionContext().enqueueAction(new DataSourceExecutionContextExecuteCallback(this, "Infragistics.Controls.DataSource.Implementation.BaseDataSource.DoRefreshInternal") {
                public void invoke() {
                    doRefreshInternal();
                }
            });
        }

    }


    public void doRefreshInternal() {
        if (getDeferAutoRefresh()) {
            _autoRefreshQueued = false;
            return;
        }

        if (!_autoRefreshQueued) {
            return;
        }

        _autoRefreshQueued = false;
        refreshInternal();
    }


    public void refreshInternal() {
       refreshInternalOverride();
    }


    protected void refreshInternalOverride() {
        removeAllPageRequests();
        killWorker();
        createWorker();
        //TODO: should we have this here? prob need to readd request for current page instead.
        _worker.addPageRequest(0, DataSourcePageRequestPriority.NORMAL);
    }


    public void flushAutoRefresh() {
        doRefreshInternal();
    }


    public void refresh() {
        refreshInternal();
    }

    private boolean _shouldDeferUpdates = false;


    @Override
    public boolean setDeferAutoRefresh(boolean value) {

        boolean oldValue = _shouldDeferUpdates;
        _shouldDeferUpdates = value;
        if (_shouldDeferUpdates != oldValue) {
            onShouldDeferAutoRefreshChanged(oldValue, value);
        }

        return value;
    }

    private void onShouldDeferAutoRefreshChanged(boolean oldValue, boolean value) {
        queueAutoRefresh();
        if (_shouldDeferUpdates && valid() && _currentSchema == null) {
            queueSchemaFetch();
        }
    }

    @Override
    public boolean getDeferAutoRefresh() {

        return _shouldDeferUpdates;

    }

    private boolean _schemaFetchQueued = false;
    public void queueSchemaFetch()
    {
        if (_schemaFetchQueued)
            {
                        return;
        }

        if (getExecutionContext() != null) {
            _schemaFetchQueued = true;
            getExecutionContext().enqueueAction(new DataSourceExecutionContextExecuteCallback() {
                @Override
                public void invoke() {
                    doSchemaFetchInternal();
                }
            });
        }
    }

    void doSchemaFetchInternal() {
        if (!_schemaFetchQueued) {
            return;
        }
        _schemaFetchQueued = false;
        schemaFetchInternal();
    }

    void schemaFetchInternal()
    {
        schemaFetchInternalOverride();
    }

    protected void schemaFetchInternalOverride() {
        if (!getDeferAutoRefresh()) {
            return;
        }
        removeAllPageRequests();
        killWorker();
        createWorker();

        addSchemaRequest();
    }

    private void addSchemaRequest() {
        _worker.addPageRequest(
                ODataVirtualDataSourceDataProviderWorker.SCHEMA_REQUEST_INDEX,
                DataSourcePageRequestPriority.HIGH
        );
    }
    

    @Override
    public boolean getNotifyUsingSourceIndexes() {
        return false;
    }

    @Override
    public int indexOfItem(Object o) {
        return -1;
    }

    @Override
    public int indexOfKey(Object[] objects) {
        return -1;
    }

    @Override
    public boolean getIsItemIndexLookupSupported() {
        return false;
    }

    @Override
    public boolean getIsKeyIndexLookupSupported() {
        return false;
    }
}
